packages/core/src/defer/triggering.ts

```diff
@@ -400,17 +407,36 @@ export async function triggerHydrationFromBlockName(
   }
 
   // Actually do the triggering and hydration of the queue of blocks
-  for (const dehydratedBlockId of hydrationQueue) {
-    await triggerResourceLoadingForHydration(dehydratedBlockId, dehydratedBlockRegistry);
-    await nextRender(injector);
-    // TODO(incremental-hydration): assert (in dev mode) that a defer block is present in the dehydrated registry
-    // at this point. If not - it means that the block has not been hydrated, for example due to different
-    // `@if` conditions on the client and the server. If we detect this case, we should also do the cleanup
-    // of all child block (promises, registry state, etc).
-    // TODO(incremental-hydration): call `rejectFn` when lDetails[DEFER_BLOCK_STATE] is `DeferBlockState.Error`.
-    blocksBeingHydrated.get(dehydratedBlockId)!.resolve();
-
-    // TODO(incremental-hydration): consider adding a wait for stability here
+  for (let i = 0; i < hydrationQueue.length; i++) {
+    const dehydratedBlockId = hydrationQueue[i];
+    const dehydratedDeferBlock = dehydratedBlockRegistry.get(dehydratedBlockId);
+
+    if (dehydratedDeferBlock != null) {
+      // trigger the block resources and await next render for hydration. This should result
+      // in the next block ɵɵdefer instruction being called and that block being added to the dehydrated registry.
+      await triggerResourceLoadingForHydration(dehydratedDeferBlock);
+      await nextRender(injector);
+
+      // if the content has changed since server rendering, we need to check for the expected block
+      // being in the registry or if errors occurred. In that case, we need to clean up the remaining expected
+      // content that won't be rendered or fetched.
+      if (deferBlockHasErrored(dehydratedDeferBlock.lView, dehydratedDeferBlock.tNode)) {
+        // Either the expected block has not yet had its ɵɵdefer instruction called or the block errored out when fetching
+        // resources. In the former case, either we're hydrating too soon or the client and server differ. In both cases,
+        // we need to clean up child content and promises.
+        const dehydratedViewsToCleanup = dehydratedDeferBlock.lContainer[DEHYDRATED_VIEWS];
+        removeDehydratedViewList(dehydratedViewsToCleanup, dehydratedDeferBlock.lContainer);
```

Comment:
nit: we can also simplify the contract here and just pass the `dehydratedDeferBlock` (calculate the rest inside of the `removeDehydratedViewsInBlock` function):
```suggestion
        removeDehydratedViewsInBlock(dehydratedDeferBlock);
```